recipes/lora_finetune_single_device.py

```diff
@@ -570,8 +588,19 @@ def _loss_step(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
         mask = batch.get("mask", None)  # shape [b, s, s]
         input_pos = batch.get("input_pos", None)  # shape [b, s]
 
-        # run model
-        logits = self._model(tokens, mask=mask, input_pos=input_pos)
+        maybe_offload_ctx = contextlib.nullcontext()
+        if self.enable_activation_offloading:
+            maybe_offload_ctx = OffloadActivations(use_streams=True)
+            noop_ctx = NoOpManager()
+            self._model.output.register_forward_pre_hook(
+                lambda *args: noop_ctx.__enter__()
+            )
+            self._model.output.register_forward_hook(
+                lambda *args: noop_ctx.__exit__(), always_call=True
+            )
```

Comment:
Just so I understand, this is our hack to ensure that output activations do not get offloaded, right? (Nice hack though.) Is this just cause those are actually too hefty, or are there other reasons for this (e.g. interaction with our chunked output layer)?

Reply:
The compute for them is super small compared to offloading them (cuz they're chunky), and they're used pretty much right away so we'd waste time moving them to CPU and waiting for them to come back. I'll add a comment for this